pkg/controller/workspace/tfc_run.go

```diff
@@ -44,28 +44,50 @@ func (t *TerraformCloudClient) CreateConfigurationVersion(workspaceID string) (*
 
 // CreateRun runs a new Terraform Cloud configuration
 func (t *TerraformCloudClient) CreateRun(workspace *v1alpha1.Workspace, terraform []byte) error {
-	configVersion, err := t.CreateConfigurationVersion(workspace.Status.WorkspaceID)
-	if err != nil {
-		return err
-	}
-
-	os.Mkdir(moduleDirectory, 0777)
-	if err := ioutil.WriteFile(configurationFilePath, terraform, 0777); err != nil {
-		return err
-	}
-
-	if err := t.UploadConfigurationFile(configVersion.UploadURL); err != nil {
-		return err
-	}
 
 	message := fmt.Sprintf("%s, apply", TerraformOperator)
 	options := tfc.RunCreateOptions{
-		Message:              &message,
-		ConfigurationVersion: configVersion,
+		Message: &message,
 		Workspace: &tfc.Workspace{
 			ID: workspace.Status.WorkspaceID,
 		},
 	}
+
+	if workspace.Spec.VCS != nil {
+		retryCnt := 0
+		foundCV := false
+		for retryCnt < 2 {
+			configVersions, err := t.Client.ConfigurationVersions.List(context.Background(), workspace.Status.WorkspaceID, tfc.ConfigurationVersionListOptions{})
+			if err != nil {
+				return err
+			}
+			if len(configVersions.Items) > 0 {
+				foundCV = true
+				break
+			}
+			retryCnt++
+			time.Sleep(100 * time.Millisecond)
```

Comment:
OK I will do that, thanks!

Reply:
They should both requeue, but the difference is the type of message left in the operator log. It could be most helpful to the user if we passed them the error returned from `t.Client.ConfigurationVersions.List`. That way TFC/TFE might tell them "I couldn't find a ConfigurationVersion for this workspace". And then they could investigate why that is. Meanwhile, the operator will keep trying to find it, which gives them a chance to log into TFC and fix the problem. (Disclaimer: I don't know much about TFC and I'm not sure what a ConfigurationVersion is offhand. heh)

I found this piece of documentation helpful yesterday when looking at requeuing options:

```
return reconcile.Result{}, nil
The reconcile process finished with no errors and does not require another pass
through the reconcile loop.

return reconcile.Result{}, err
The reconcile failed due to an error and Kubernetes should requeue it to try
again.

return reconcile.Result{Requeue: true}, nil
The reconcile did not encounter an error, but Kubernetes should requeue it to
run for another iteration.

return reconcile.Result{RequeueAfter: time.Second*5}, nil
Similar to the previous result, but this will wait for the specified amount of time
before requeuing the request. This approach is useful when there are multiple
steps that must run serially, but may take some time to complete. For example, if
a backend service needs a running database prior to starting, the reconcile can be
requeued with a delay to give the database time to start. Once the database is run‐
ning, the Operator does not requeue the reconcile request, and the rest of the
steps continue.
```
https://www.redhat.com/cms/managed-files/cl-oreilly-kubernetes-operators-ebook-f21452-202001-en_2.pdf

Reply:
The ConfigurationVersion object is not immediately available when using a VCS backed workspace, which is why I put the retry in the first place. 

Based on the description above I think RequeueAfter is the best course of action.